Combo

noahadams opened the pull request:
Integrates use of combo with localStorage based caching, and some tests of said feature

Commit:
Comobo Service.

noahadams committed:
ensure double assignment on clear cache so closed over resources stays the same object as Mobify.combo.resources

noahadams committed:
doctype on the test

noahadams committed:
unit test of cache rehydration

noahadams committed:
Merge branch 'refs/heads/master' into combo

noahadams committed:
move serialized variable into proper scope

noahadams committed:
"happy path" unit tests for rehydrate and dehydrate

noahadams committed:
fix initialization of lruTime

noahadams committed:
add a unit test for cache eviction

noahadams committed:
another unit test for cache dehydration/eviction

noahadams committed:
backed off waiting timeouts to succeed on slower hardware

Commit:
Fix combo for use with the Mobify Client.

- Fix dustjs syntax.
- Use Mobify.$
- Formatting fixes.

noahadams committed:
remove debug output in test

noahadams committed:
refer to START_OF_TIME

noahadams committed:
add staleness checking tests and stale eviction tests

noahadams committed:
fix to properly handle use of date header in computing a freshness lifetime for resources with a  cache-control header

noahadams committed:
add tests for cache expiry and eviction of resources with date and cache control headers

noahadams committed:
add a couple more assertions for good measure

noahadams committed:
Merge remote-tracking branch 'origin/combo-fixes' into combo

noahadams committed:
catch an additional renaming of k to key

noahadams committed:
renamed dehydrateCache() to storeCache() and rehydrateCache() to loadCache()

noahadams committed:
Merge remote-tracking branch 'origin/master' into combo

noahadams committed:
obsoleted

vmarta committed:
Minor fixes for combo

Fixed wrong names for variable/function and initialized some variable.

The async function is still not working well, though. I can see the
following script as a result:

<script type="text/javascript">Mobify.combo.loadAsync('...');<br>Mobify.combo...

For some reason, there's <br> within the script.

vmarta committed:
Noah's fixes for combo async

vmarta committed:
Firefox fix for loadSync

Possibly related to this bug:
https://github.com/mobify/zepto/commit/f0f4c12f664716be51b1d8423e044ddcf97a2ca4

Seems like Firefox toString() doesn't like to see </script>. So I had to
break it up into 2 strings.

Screenshots of problem:

- top of page:
  https://img.skitch.com/20120613-c85ci7y5m9t1dtu86f5cb1kf14.jpg
- scrolling down a bit:
  https://img.skitch.com/20120613-r8brsh5bdbcg5mn4xqmeqxjdhp.jpg

noahadams committed:
add pure javascript url absolutification to util.

noahadams committed:
Merge remote-tracking branch 'origin/url-absolutification' into combo

noahadams committed:
add some initial documentation in markdown for combo

noahadams committed:
housekeeping around change from dehydrate/rehydrate to store/load

noahadams committed:
rename comboScriptSync to combineScripts

noahadams committed:
update tests for new function name

noahadams committed:
branch on possible async parameter to combineScripts

noahadams committed:
fix for async script loading

noahadams committed:
remove debugging info

noahadams committed:
tests for both sync and async combo usage

noahadams committed:
update doc with renamed function

noahadams committed:
more explicit conditions

noahadams committed:
more comments

noahadams committed:
Revert "Merge remote-tracking branch 'origin/url-absolutification' into combo"

This reverts commit 4a5bec76fd3d984d32dfcedf3849b0533d50ac44, reversing
changes made to b4f20edda01e4b010c77c36b05aa020cb5fa3d9a.

noahadams committed:
create proper bootstrap tag for all already cached case in async branch of combineScripts()

johnboxall commented:
`$.fn.combineScripts` should probably handle the cases where: No scripts were found / No external scripts were found.

Also, let's say `localStorage` isn't available - are we looking at a performance penalty? If so, if persistence isn't available, maybe we should just return the scripts we find in order.

rrjamie commented on api/combo-caching.js:
```diff
@@ -164,12 +175,15 @@ var ccPublic = /^\s*public\s*$/
+            //DEBUG
+            // console.log('urls: ' + JSON.stringify(urls));
+            // console.log('notCachedUrls: ' + JSON.stringify(notCachedUrls));
+
```
Remove // Debug comments if possible.

Commit:
Combo refactor.

Separate code into three modules:

1. httpCache - HTTP 1.1 cache that uses localStorage for persistence.
2. httpCache.utils - Set of functions for handling caching headers.
3. Combo - Set of function for using the Combo api.

Commit:
Combo: Fix a failing test.

Commit:
Combo: Properly reference $ inside httpCache. `combo.getURL` always returns the same URL for the same set of URLs.

Commit:
Merge branch 'refs/heads/master' into combo

Conflicts:
	konf/base.konf

Commit:
Combo: Remove unused caching logic.

rrjamie commented:
What's the status of this?

johnboxall commented:
@rrjamie the latest code is awaiting review to be merged.

rrjamie commented on api/combo.js:
```diff
@@ -0,0 +1,330 @@
+            // If `expires` is present, we are stale if we are older.
+            if (expires) {
+                return now > Date.parse(expires);
+            }
```
Is this what we want? For non-parsables to never be stale:

```
 >>> 1 > Date.parse("Barf Barf Barf")
 false
```

rrjamie commented on api/combo.js:
```diff
@@ -0,0 +1,330 @@
+            }
+
+            // Otherwise, we are stale.
+            return false;
```
If we are stale, return true?

Commit:
Combo bug fixes.

- Switch testing server to connect
- Handle the case where all combo scripts are cached
- If the date header is malformed, ignore max-age
- If no caching headers are present, a resource is stale

Commit:
Combo: invalid expires is stale.
